lib/recurly/client.rb

```diff
@@ -118,18 +124,67 @@ def delete(path, **options)
     # @return [Logger]
     attr_reader :logger
 
-    def run_request(request, headers)
-      read_headers @conn.run_request(request.method, request.path, request.body, headers)
+    @connection_pool = Recurly::ConnectionPool.new
+
+    class << self
+      # @return [Recurly::ConnectionPool]
+      attr_accessor :connection_pool
     end
 
-    def handle_response!(request, faraday_resp)
-      response = HTTP::Response.new(faraday_resp, request)
-      raise_api_error!(response) unless (200...300).include?(response.status)
+    def run_request(request, options = {})
+      self.class.connection_pool.with_connection do |http|
+        set_http_options(http, options)
+
+        retries = 0
+
+        begin
+          http.start unless http.started?
+          http.request(request)
+        rescue EOFError, Errno::ECONNREFUSED, Errno::ECONNRESET, Errno::EHOSTUNREACH, Errno::ECONNABORTED, Errno::EPIPE, Errno::ETIMEDOUT, Net::OpenTimeout => ex
+          retries += 1
+          if retries < MAX_RETRIES
+            retry
+          end
+
+          if ex.kind_of?(Net::OpenTimeout) || ex.kind_of?(Errno::ETIMEDOUT)
+            raise Recurly::Errors::TimeoutError, "Request timed out"
+          end
+
+          raise Recurly::Errors::ConnectionFailedError, "Failed to connect to Recurly: #{ex.message}"
+        rescue Net::ReadTimeout, Timeout::Error
+          raise Recurly::Errors::TimeoutError, "Request timed out"
+        rescue OpenSSL::SSL::SSLError => ex
+          raise Recurly::Errors::SSLError, ex.message
+        end
```

Comment:
Should we have a default rescue here? Also should we catch and translate SocketError as well?
```
Traceback (most recent call last):
	13: from scripts/v2019-10-10/ruby/create_account.rb:34:in `<main>'
	12: from /app/clients/ruby/lib/recurly/client/operations.rb:147:in `create_account'
	11: from /app/clients/ruby/lib/recurly/client.rb:90:in `post'
	10: from /app/clients/ruby/lib/recurly/client.rb:135:in `run_request'
	 9: from /app/clients/ruby/lib/recurly/connection_pool.rb:19:in `with_connection'
	 8: from /app/clients/ruby/lib/recurly/client.rb:141:in `block in run_request'
	 7: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:925:in `start'
	 6: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
	 5: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:945:in `connect'
	 4: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:103:in `timeout'
	 3: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
	 2: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:947:in `block in connect'
	 1: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:947:in `open'
/usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:947:in `initialize': getaddrinfo: Name or service not known (SocketError)
	12: from scripts/v2019-10-10/ruby/create_account.rb:34:in `<main>'
	11: from /app/clients/ruby/lib/recurly/client/operations.rb:147:in `create_account'
	10: from /app/clients/ruby/lib/recurly/client.rb:90:in `post'
	 9: from /app/clients/ruby/lib/recurly/client.rb:135:in `run_request'
	 8: from /app/clients/ruby/lib/recurly/connection_pool.rb:19:in `with_connection'
	 7: from /app/clients/ruby/lib/recurly/client.rb:141:in `block in run_request'
	 6: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:925:in `start'
	 5: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:930:in `do_start'
	 4: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:945:in `connect'
	 3: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:103:in `timeout'
	 2: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/timeout.rb:93:in `block in timeout'
	 1: from /usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:946:in `block in connect'
/usr/local/rbenv/versions/2.6.3/lib/ruby/2.6.0/net/http.rb:949:in `rescue in block in connect': Failed to open TCP connection to v3.recurly.con:443 (getaddrinfo: Name or service not known) (SocketError)
```

Reply:
👍